main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala

```diff
@@ -0,0 +1,489 @@
+/*
+ * Copyright 2024 Herluf Baggesen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.language.errors.Parser2Error
+
+/**
+ * A Concrete syntax tree representing the textual content of a source file.
+ * Note the difference compared to an __abstract__ syntax tree,
+ * which aims to represent the semantic meaning of a program rather than the text itself.
+ *
+ * The tree is built of nodes that hold a [[TreeKind]] and zero or more children.
+ * Each child is either a [[Token]] or another node.
+ * [[SyntaxTree]] is intentionally kept very flexible in definition, so it can encode faulty syntax.
+ * Beware that this means that it gives few guarantees.
+ * There is no guarantee on the amount or presence of children of a certain kind for instance.
+ */
+object SyntaxTree {
+
+  /**
+   * A node in a [[SyntaxTree]]
+   *
+   * @param kind     The kind of the node.
+   * @param loc      The location that the node spans in the source file.
+   * @param children The children of the node.
+   */
+  case class Tree(kind: TreeKind, var loc: SourceLocation, var children: Array[Child])
+
+  sealed trait Child
+
+  /**
+   * A child in a [[SyntaxTree]].
+   * It holds a [[Token]] or another [[SyntaxTree.Tree]].
+   */
+  object Child {
+    /**
+     * A [[SyntaxTree]] child holding a [[Token]].
+     */
+    case class NodeToken(token: Token) extends Child
+
+    /**
+     * A [[SyntaxTree]] child holding a nested [[SyntaxTree.Tree]]
+     */
+    case class NodeTree(tree: SyntaxTree.Tree) extends Child
+  }
+
+
+  /**
+   * A common super-type for [[TreeKind]]s
+   */
+  sealed trait TreeKind
+
+  /**
+   * Different kinds of syntax nodes in a [[SyntaxTree]].
+   * The only error kind that holds data is the special [[TreeKind.ErrorTree]].
+   */
+  object TreeKind {
+    /**
+     * A special error kind wrapping a [[Parser2Error]].
+     */
+    case class ErrorTree(error: Parser2Error) extends TreeKind
+
+    case object AnnotationList extends TreeKind
```

Comment:
Intuitively, why is there no Annotation kind? Is it because it is just a token?

Reply:
Yep, the `TokenKind` is enough to recognize an annotation. If there is a child with a kind that is not an annotation, that's a parser error!